src/Components/Components/src/NavigationManager.cs

```diff
@@ -104,6 +116,19 @@ public void NavigateTo(string uri, bool forceLoad = false)
         /// <param name="forceLoad">If true, bypasses client-side routing and forces the browser to load the new page from the server, whether or not the URI would normally be handled by the client-side router.</param>
         protected abstract void NavigateToCore(string uri, bool forceLoad);
```

Comment:
This does seem a bit odd to me,
We now have an abstract function which descendant classes are forced to implement,
But we would like them to override the function below to use the new NavigationOptions. 

A non breaking change would be to make this function virtual instead of abstract (which is non breaking) (And make it throw a NotImplemented exception)
And add an obsolete attribute to it, explaining that descendants should override NavigateToCore(string uri, NavigationOptions options) to implement routing.

This might be acceptable as typical users would not subclass NavigationManager,

Reply:
> A non breaking change would be to make this function virtual instead of abstract (which is non breaking) (And make it throw a NotImplemented exception)

I'm afraid I don't follow how the suggestion design would really work. Currently, `NavigationManager.NavigateTo` already calls `NavigateToCore(string, bool)`. Are you proposing we change `NavigateTo` to call the `(string, NavigationOptions)` overload instead?

 * If yes, then we need the `(string, NavigationOptions)` overload to have useful default behavior. And the only thing it could do is call through to the `(string, bool)` overload since that's what's implemented on existing subclasses. But then we need `(string, bool)` not to throw, since we're calling it ourselves.
 * If no, then we we're calling the `(string, bool)` overload ourselves and of course need it not to throw.

So I don't see how in either case we could have the `(string, bool)` overload throw.

> This might be acceptable as typical users would not subclass NavigationManager,

Perhaps you're suggesting we just have a breaking change. If that's the suggestion, we'd be better off just taking a source-breaking change and changing the `(string, bool)` overload to `(string, NavigationOptions)` without adding a second overload. But our goal here is not to have a breaking change.

Reply:
Sorry if I wasn't clear, 
 I am talking about the ``Navigationmanager.NavigateToCore`` function,

It used to be:
```cs
protected abstract void NavigateToCore(string uri, bool forceLoad);
````
But this now seems odd, to have an abstract function, that descendants need to override. when we are introducing a new NavigateToCore function with the new NavigationOptions

to make sure we don't introduce a breaking change I propose to change the abstract function to a virtual one.
And mark it as Obsolete

```cs
[Obsolete("Override NavigateToCore(string uri, NavigationOptions options) instead")]
protected virtual void NavigateToCore(string uri, bool forceLoad) => throw new NotImplementedException();

protected virtual void NavigateToCore(string uri, NavigationOptions options)
{
  //We call NavigateToCore(string uri, bool forceLoad) so to not introduce a breaking change.
  //derived classes should override this function to implement routing
  NavigateToCore(uri, options.ForceLoad);
}
```

This way, all users that use custom NavigationManager Descendants will continue to work. (Although I don't think there will be a lot of Custom NavigationManagers (Except maybe mocked ones) as it's not possible to register ones own NavigationManager ). But even then these descendants will still function as before. (If they override `` void NavigateToCore(string uri, bool forceLoad)`` As they have to because it was marked abstract,  their override will still be called.

And the Obsolete message will only be visible for people who create their own NavigationManager descendant, not the majority of users.

And just for completeness,
The Navigationmanager.NavigateTo original and overloads are:

```cs
public void NavigateTo(string uri, bool forceLoad = false)
{
  AssertInitialized();
  NavigateToCore(uri, new NavigationOptions { ForceLoad = forceLoad });
}

public void NavigateTo(string uri, NavigationOptions options)
{
  AssertInitialized();
  NavigateToCore(uri, options);
}
```

Reply:
Updated PR using https://github.com/dotnet/roslyn/blob/master/docs/Adding%20Optional%20Parameters%20in%20Public%20API.md